### ISSUE TITLE
Redis dictionary with direct layout has strange exceptions on some queries and never recovers

### ISSUE THREAD

**Issue by shawel** [ClickHouse/ClickHouse#12580]
I get this on the following queries and the only thing i changed is the limit on the query
` DB::Exception: Assertion violation: _output in file "../contrib/poco/Redis/src/Client.cpp `

First query is successful.

<img width="1528" alt="Screen Shot 2020-07-19 at 9 34 06 AM" src="https://user-images.githubusercontent.com/625370/87876151-186f7a80-c9a4-11ea-99c0-0ebf07d1858d.png">

Here is the log
<img width="1609" alt="Screen Shot 2020-07-19 at 9 36 26 AM" src="https://user-images.githubusercontent.com/625370/87876167-32a95880-c9a4-11ea-87a8-453c955a3952.png">

If i repeat the first successful query it throws the same error. The only way to recover is to drop and recreate the dictionary

Here is the dictionary 
```
 CREATE DICTIONARY default.dict_redis
(
    `ID` UInt64,
    `value` String DEFAULT '0'
)
PRIMARY KEY ID
SOURCE(REDIS(HOST 'redis-svc.misc.svc' PORT 6379 STORAGE_TYPE 'simple' DB_INDEX 0))
LAYOUT(DIRECT()) 
```

**Comment by CurtizJ** [comment]
I found a bug, but it can appear only if query failed to connect to Redis and it fails with exception `Lost connection to Redis server`. After that all queries to dictionary fail with `DB::Exception: Assertion violation: _output in file`. Are you sure, that there was no failed queries beetween your runs?

**Comment by shawel** [comment]
Hey @CurtizJ 

I am not sure if it is a connection problem. It is a container in the same pod on Kubernetes so it is almost as if it is on localhost. and i don't see a pod failure. And that final test fails consistently.  it reproduces without fail.

**Comment by CurtizJ** [comment]
I didn't reproduce your case, but I investigated Poco redis library and fixed in #13082 the only case, that could lead to this kind of errors. So, mostly likely it is already fixed.
